### DISCUSSION TITLE
Other Setting Implementation

### DISCUSSION THREAD

**Discussion by disruptivepatternmaterial** [echavet/MitsubishiCN105ESPHome#69] [General]
So the calling pattern on the serial seems to be that the kumo sends an empty sentence of the type it is interested in. Then the split sends back all the setting in that set. Then when kumo adapter wants to change one setting it sends back all of them in a command.

    From Kumo FC 42 01 30 10 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5D 
    From Pump FC 62 01 30 10 20 04 08 0C 10 14 18 3C 40 44 48 4C 50 54 58 00 99

The same thing happens for the 4 commands.

So I am wondering where a place to start is @echavet ?

Since I don't know what they all mean, I cannot exactly follow the map/byte pattern that comes from the [swicago](https://github.com/SwiCago/HeatPump) project.

It seems like periodically we need to send the empty sentence, and store it as is until we have a map. Then create a new setting for Thermal Fan (a place to start) and rebuild the sentence. So it will have to be a tad messy until maybe I can figure out what the other bytes mean.

I understood what you were up to @echavet until your last major refactor :P so I will have to spend time reabsorbing your pattern! But I think I can just insert this in your "loop" that is checking temp and others...

As some point I am going to give it a go as a POC then once we have a pattern we can iteratively add settings. But there are so many I cannot figure out if there are 4 with 15 each = 60! WTHeck could they be?

**Comment by disruptivepatternmaterial** [comment]
1. CN105Climate::getExtraSetting() {} in hp_writings.cpp - sends the empty sentences.
2. CN105Climate::getExtraSettingsFromResponsePacket() {} in hp_readings.cpp - processes them
3. CN105Climate::setExtraSettings() {} - sends the empty sentences - needs to pull know setting from the get, and adjust with wanted changes from user input

Seems that makes sense as the implantation order.

I am not sure what needs to be passed at this moment, nor am I certain about how to handle all 4 types. My hunch is to handle them in the same methods but 4 times each, but I am thinking I'll try one first.

**Reply by echavet** [reply]
Yes, it could be like that. However, it should first begin with experimentation, which might later lead to integration into the current cycle.
It could also lead us to reevaluate the request/response mechanism!
I need to think about what you have found.

Do you have other examples formatted like that:

```
From Kumo FC 42 01 30 10 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5D 
From Pump FC 62 01 30 10 20 04 08 0C 10 14 18 3C 40 44 48 4C 50 54 58 00 99
```

**Comment by echavet** [comment]
What you have discovered could be major breakthrough. I'm not sure yet what we can do with it, but I will think about it.

I made **two changes** in my last refactoring. It was initiated by the desire to circumvent the difficulty of uncertainty regarding the air conditioner's ACK (following a request to change operating parameters or an update of the outdoor temperature). So, I wanted to make the mechanism more optimistic. 
The second aim was to limit concurrent accesses.

This means that:
1. **I assumed that the air conditioner received all requests**, and I no longer wait for its response to modify the state of the graphical component.
2. **I simplified potentially concurrent exchanges** (back and forth) by forcing a **sequencing of requests/responses** (which is called **Cycle**).
The idea is that the main loop oscillates between 3 functions with a priority:
   1. Reading
   2. Order (wantedSettings)
   3. Request cycles

- For the readings, nothing has changed. If the air conditioner speaks to us, we listen, and it's the absolute priority.
- The wantedSettings, it's the second priority: if the user asks for something, they take over.
- Then comes the query cycle. It unfolds naturally:
It starts with:
   - 1: Cycle Start
   then follows sequential request/response. That is, I send the first request and then wait for the response. When I receive it, I send the next request, and then wait for the response again. "Waiting" is, of course, asynchronous.

   - 2a: Request:       Sending settings request (0x02)
   - 2b: Response:   Receiving settings response
   - 3a: Request:      Sending room °C request (0x03)
   - 3b: Response:  Receiving room °C response
   - 4a: Request:     Sending status request (0x06)
   - 4b: Response:  Receiving status response (0x06)
   - 5a: Request:     Sending power request (0x09)
   - 5b: Response:  Receiving Power/Standby response
   - 6: Cycle Stop

**Comment by echavet** [comment]
From Kumo FC 42 01 30 10 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5D 
From Pump FC 62 01 30 10 20 04 08 0C 10 14 18 3C 40 44 48 4C 50 54 58 00 99

FC ok
62 ok (req infos)
01 30 ok (header is always like that)
10 ok: 16 data bytes
20 ok: we don't use it but SwiCago has identified this as heatpump functions.

I've never tested this (despite I have kept the code in my component). It would be a big progress to use this.

Actually I think the only way to decode these bytes is to have many different types of unit and to compare all the different responses from each model. 
As it is a "functions" request, my guess is that it shouldn't find his place in the cycle loop but it should come next to the identification success, before the request infos cycle starts. Does the kumo repeat theses requests constantly in the loop process ? Or does it only request functions after the connection ?


```cpp
void CN105Climate::processCommand() {
    switch (this->command) {
    case 0x61:  /* last update was successful */
        this->hpPacketDebug(this->storedInputData, this->bytesRead + 1, "Update-ACK");
        this->updateSuccess();
        break;

    case 0x62:  /* packet contains data (room °C, settings, timer, status, or functions...)*/
        this->getDataFromResponsePacket();
        break;
    case 0x7a:
        ESP_LOGI(TAG, "--> Heatpump did reply: connection success! <--");
        // ======================================
        // here could come the heatpump functions requests
        // But the 10s timeout in the sendFirstConnectionPacket() should probably be increased
        // and the comming "this->isHeatpumpConnected_ = true;" should be deferred after the hp response
        // ======================================
                
        this->isHeatpumpConnected_ = true;     //should be deferred after the hp response as mentionned 

        // let's say that the last complete cycle was over now        
        this->loopCycle.lastCompleteCycleMs = CUSTOM_MILLIS;
        this->currentSettings.resetSettings();      // each time we connect, we need to reset current setting to force a complete sync with ha component state and receievdSettings 
        break;
    default:
        break;
    }
}
```

**Reply by disruptivepatternmaterial** [reply]
sorry...life was distracting me :) I was traveling and thinking about this as I sat in airports...

you are 100% right, the kumo seems to not ask for these in a cycle, it does when you enter the settings pages in the app or seemingly when it starts up. I can try this over time and see if it does on some longer cycle.

the thing is that is stores the ENTIRE thing and sends it ALL, so that is critical to the process.

This was done on a MSZ-GL09NA, but I'd bet money you are right - to do this right we have to make a library of units which is annoying.

Like I said before, I plan to move this to another unit which is of another type and has hyper heat, isee, etc... then we can see what that ones says (all my units are MSZ-GLXXNA except that one)

As soon as I am back in the office where that other unit is, I'll give it a go!
